msprime/species_tree.py

```diff
@@ -0,0 +1,489 @@
+    Leafs must be named. An example for an accepted tree string in
+    Newick format is:
+    (((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)
+    The tree string can end with a semi-colon, but this is not required.
+
+    - An estimate of the effective population size Ne should be
+        specified.
+    - If and only if the branch lengths are not in units of
+        generations, the generation time should be specified.
+    """
+
+    # Make sure a species tree is specified.
+    if species_tree is None:
+        raise ValueError("A species tree must be specified.")
+
+    # Make sure that branch length units are either "myr", "yr", or "gen".
+    allowed_branch_lenth_units = ["myr", "yr", "gen"]
+    if branch_length_units not in allowed_branch_lenth_units:
+        err = 'The specified units for branch lengths ('
+        err += '"{}") are not accepted. '.format(branch_length_units)
+        err += 'Accepted units are "myr" (millions of years), "yr" (years), '
+        err += 'and "gen" (generations).'
+        raise ValueError(err)
+
+    # Make sure that the population size is either None or positive.
+    if Ne is not None and Ne <= 0:
+        raise ValueError("Population size Ne must be > 0")
+
+    # Make sure that the generation time is either None or positive.
+    if generation_time is not None and generation_time <= 0:
+        raise ValueError("Generation time must be > 0")
+
+    # Make sure that the generation time is specified if and only if
+    # branch lengths are not in units of generations.
+    if branch_length_units == "gen":
+        if generation_time is not None:
+            err = 'With branch lengths in units of generations ("gen"), '
+            err += 'a generation time should not be specified additionally.'
+            raise ValueError(err)
+    else:
+        if generation_time is None:
+            err = 'With branch lengths in units of '
+            err += '"{}", a generation time must be '.format(branch_length_units)
+            err += 'specified additionally.'
+            raise ValueError(err)
+
+    # Make sure that a population size is specified.
+    if Ne is None:
+        raise ValueError("Ne should be specified.")
+
+    # Get the number of generations per branch length unit.
+    generations_per_branch_length_unit = get_generations_per_branch_length_unit(
+        branch_length_units, generation_time
+        )
+
+    # Read the input file.
+    species_tree_lines = species_tree.splitlines(False)
+    if len(species_tree_lines) > 1:
+        raise ValueError("The species tree has multiple lines.")
+    species_tree_string = get_species_tree_string(species_tree_lines[0])
+
+    # Parse the newick tree string.
+    root = newick.loads(species_tree_string)[0]
+
+    # Remove nodes that have only a single child if there should be any.
+    root.remove_redundant_nodes()
+
+    # Resolve polytomies by injecting zero-length branches.
+    root.resolve_polytomies()
```

Comment:
It seems `resolve_polytomies()` isn't exactly doing what I thought it would. Not sure if the ancestor property is not getting updated as it should:
```
import newick

def print_node_heights(tree):
    root = tree
    # Set node depths (distances from root)
    max_depth = 0
    for node in root.walk():
        depth = 0
        moving_node = node
        while moving_node.ancestor:
            depth += moving_node.length
            moving_node = moving_node.ancestor
        node.depth = depth
        if depth > max_depth:
            max_depth = depth
    # Set node heights (distances from present).
    for node in root.walk():
        node.height = max_depth - node.depth
        print(node.height)

root = newick.loads("((human:1,chimpanzee:1,gorilla:1):2,orangutan:3)")[0]
print_node_heights(root)
print()
root.resolve_polytomies()
print_node_heights(root)
```
Anyway, the best way to avoid this issue may be not to use resolve_polytomies, so I'll follow your advice and remove the dependence on bifurcating trees (at least for the newick parsing, not for the starbeast parsing though).

Reply:
The output of the above code is:
```
3.0
1.0
0.0
0.0
0.0
0.0

```(before calling resolve_polytomies)
and 
```
3.0
1.0
0.0
3.0
2.0
2.0
0.0

```
(after calling it). So somehow nodes end up having a height of  2 (which is not in the input tree) and fewer have a height of 0 (even though four nodes are extant).

Reply:
Ok, resolve_polytomies() is no longer used, and neither is remove_redundant_nodes(). Instead the code now works with trees that have nodes with any number of descendants.